packages/@aws-cdk/aws-redshift/lib/parameter-group.ts

```diff
@@ -62,17 +63,46 @@ export class ClusterParameterGroup extends ClusterParameterGroupBase {
    */
   public readonly clusterParameterGroupName: string;
 
+  /**
+   * The parameters in the parameter group
+  */
+  readonly parameters: { [name: string]: string };
+
+  /**
+   * The underlying CfnClusterParameterGroup
+  */
+  private readonly resource: CfnClusterParameterGroup;
+
   constructor(scope: Construct, id: string, props: ClusterParameterGroupProps) {
     super(scope, id);
-
-    const resource = new CfnClusterParameterGroup(this, 'Resource', {
+    this.parameters = props.parameters;
+    this.resource = new CfnClusterParameterGroup(this, 'Resource', {
       description: props.description || 'Cluster parameter group for family redshift-1.0',
       parameterGroupFamily: 'redshift-1.0',
-      parameters: Object.entries(props.parameters).map(([name, value]) => {
-        return { parameterName: name, parameterValue: value };
-      }),
+      parameters: this.parseParameters(),
     });
 
-    this.clusterParameterGroupName = resource.ref;
+    this.clusterParameterGroupName = this.resource.ref;
+  }
+  private parseParameters(): any {
+    return Object.entries(this.parameters).map(([name, value]) => {
+      return { parameterName: name, parameterValue: value };
+    });
+  }
+
+  /**
+   * Adds a parameter to the parameter group
+   *
+   * @param name the parameter name
+   * @param value the parameter name
+   */
+  public addParameter(name: string, value: string): void {
+    const existingValue = Object.entries(this.parameters).find(([key, _]) => key === name)?.[1];
+    if (existingValue === undefined) {
+      this.parameters[name] = value;
+      this.resource.parameters = this.parseParameters();
+    } else if (existingValue !== value) {
+      throw new Error(`The parameter group already contains the parameter "${name}", but with a different value (Given: ${value}, Existing: ${existingValue}).`);
```

Comment:
Parameter groups don't have names, but I agree that it makes sense to identify it somehow , so maybe it makes sense to use `Names.uniqueResourceName` here

```ts
  throw new Error(`The parameter group id "${Names.uniqueResourceName(this, {})}" already contains the parameter "${name}", but with a different value (Given: ${value}, Existing: ${existingValue}).`);
```

Reply:
See my most recent comment below.

Reply:
In this case it's [set to](https://github.com/aws/aws-cdk/pull/20944/files/36217333f8b33a2b309c28adb0bdf04b6bb84a9e#diff-58ab3674f62ed150088a23634fecaa1654c40f1911a0fb271935e0fc127c509dR85) the Instrinsic `Ref` of `CfnClusterParameterGroup`, so the error message would contain a `Token`.

Considering that this is runtime error, I think using `Names.uniqueResourceName` (along with the stack trace) should be helpful enough to locate the problematic parameter group

Reply:
This would be providing the customer with inaccurate information. Sure, it's an approximation, but it's not actually the name of the resource. I suppose the stack trace will have to provide enough information in conjunction with the key value pair they're trying to add.

Reply:
Reverted back to this message in the latest commit